pkg/monitoring/metrics/forklift-controller/migration_metrics.go

```diff
@@ -67,53 +68,63 @@ func RecordMigrationMetrics(c client.Client) {
 
 				provider := sourceProvider.Type().String()
-
-				if m.Status.HasCondition(Succeeded) {
-					key = fmt.Sprintf("%s|%s|%s|%s|%s", Succeeded, provider, mode, target, string(plan.UID))
-					counterMap[key]++
-
-					startTime := m.Status.Started.Time
-					endTime := m.Status.Completed.Time
-					duration := endTime.Sub(startTime).Seconds()
-
-					var totalDataTransferred float64
-					for _, vm := range m.Status.VMs {
-						for _, step := range vm.Pipeline {
-							if step.Name == "DiskTransferV2v" || step.Name == "DiskTransfer" {
-								for _, task := range step.Tasks {
-									totalDataTransferred += float64(task.Progress.Completed) * 1024 * 1024 // convert to Bytes
-								}
-							}
-						}
-					}
-
-					// Set the metrics for duration and data transferred and update the map for scaned migration
-					if _, exists := processedSucceededMigrations[string(m.UID)]; !exists {
-						migrationDurationGauge.With(prometheus.Labels{"provider": provider, "mode": mode, "target": target, "plan": string(plan.UID)}).Set(duration)
-						migrationDurationHistogram.With(prometheus.Labels{"provider": provider, "mode": mode, "target": target}).Observe(duration)
-						dataTransferredGauge.With(prometheus.Labels{"provider": provider, "mode": mode, "target": target, "plan": string(plan.UID)}).Set(totalDataTransferred)
-						processedSucceededMigrations[string(m.UID)] = struct{}{}
-					}
-				}
-				if m.Status.HasCondition(Failed) {
-					key = fmt.Sprintf("%s|%s|%s|%s|%s", Failed, provider, mode, target, string(plan.UID))
-					counterMap[key]++
-				}
-				if m.Status.HasCondition(Executing) {
-					key = fmt.Sprintf("%s|%s|%s|%s|%s", Executing, provider, mode, target, string(plan.UID))
-					counterMap[key]++
-				}
-				if m.Status.HasCondition(Canceled) {
-					key = fmt.Sprintf("%s|%s|%s|%s|%s", Canceled, provider, mode, target, string(plan.UID))
-					counterMap[key]++
-				}
+				processMigration(m, provider, mode, target, string(plan.UID))
 			}
+		}
+	}()
+}
 
-			for key, value := range counterMap {
-				parts := strings.Split(key, "|")
-				migrationStatusGauge.With(prometheus.Labels{"status": parts[0], "provider": parts[1], "mode": parts[2], "target": parts[3]}).Set(value)
-				migrationPlanCorrelationStatusGauge.With(prometheus.Labels{"status": parts[0], "provider": parts[1], "mode": parts[2], "target": parts[3], "plan": parts[4]}).Set(value)
+func processMigration(migration api.Migration, provider, mode, target, planUID string) {
+	var (
+		processedMigrations map[string]struct{}
+		status              string
+	)
+
+	if migration.Status.HasCondition(Succeeded) {
+		processedMigrations = processedSucceededMigrations
+		status = Succeeded
+	} else if migration.Status.HasCondition(Failed) {
+		processedMigrations = processedFailedMigrations
+		status = Failed
+	} else if migration.Status.HasCondition(Canceled) {
+		processedMigrations = processedCanceledMigrations
+		status = Canceled
+	} else {
+		// we hit an undesired status
```

Comment:
I see what you meant but "undesired" doesn't sound like the right term (e.g., the migration may just be running)
I'd simply right "otherwise, there's nothing to do with the current state of the migration"